src/adaptors/map.rs

```diff
@@ -0,0 +1,120 @@
+use std::iter::FromIterator;
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct MapSpecialCase<I, F> {
```

Comment:
```suggestion
pub struct MapSpecialCase<I, F: ?Sized> {
```

This way unsized “special functions” can be used, including trait objects.

Reply:
Thanks for this nice idea. In general, I am leaning to incorporate changes that generalize functionality.

However, I'd like to postpone this one and - if we decide to accept `?Sized` functions - do it uniformly throughout itertools. I think my changes did not change behavior in this regard, so I hope postponing is ok.

@pthariensflame As I'm no expert in this area: Could you showcase a situation that requires `F: ?Sized`?

Reply:
Sure!  If we have an unsized trait object `dyn Fn(X) -> Y`, then we get an unsized `MapSpecialCaseFnOk<dyn Fn(X) -> Y>`, which is then your example of an unsized `F`.

Reply:
> However, I'd like to postpone this one and - if we decide to accept ?Sized functions - do it uniformly throughout itertools. I think my changes did not change behavior in this regard, so I hope postponing is ok.

Fine by me!

Reply:
@pthariensflame I'm curious about the actual use-case `dyn Fn` might appear in. I've never needed to dynamically dispatched function objects before.

Reply:
Yeah, I'm fine with waiting too.  This isn't essential functionality, just mildly helpful at best.

As far as `dyn Fn`, I can't point to any specific code, but it might come up when abstracting over dynamically swappable actions.  For example:  imagine a pipeline of scriptable blocks that the user can specify in a GUI, and it gets compiled down to functions that invoke the scripts and munge the values flowing through as necessary.  The whole pipeline is then composed and boxed up as a `dyn Fn` so that it can be altered on the fly as the application runs.